### PR TITLE
Implement server-side remote error logging

### LINKED ISSUE #49287: [Remote Error Logging] Implement server-side PHP remote error logging
## Description

Implement server-side PHP remote error logging for WooCommerce. To capture and log errors, use existing "fatal error" logging/handling mechanisms in WP and WC that run [on shutdown hook](https://developer.wordpress.org/reference/hooks/shutdown/). To address performance concerns, use the async scheduler to send errors, ensuring that the logging system does not negatively impact user experience.

Besides, throttling should be implemented to control the rate at which errors are logged to prevent overwhelming the logging system.

Logstash API: fbhepr%2Skers%2Sjcpbz%2Schoyvp.ncv%2Serfg%2Sjcpbz%2Qwfba%2Qraqcbvagf%2Spynff.jcpbz%2Qwfba%2Qncv%2Qybtfgnfu%2Qraqcbvag.cuc%3Se%3Qnr27n451-og

WPCOM WooCommerce Logstash handler: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Sjbbpbzzrepr%2Svapyhqrf%2Spynff%2Qjp%2Qybtfgnfu%2Qunaqyre.cuc%3Se%3Q8p4146o2%2396-og

## Accept Criteria

1. Implement remote error logging by hooking into the shutdown action or by registering a shutdown function with PHP.
2. Capture and log the last error that occurred before the shutdown. 
3. Use the async scheduler to send errors to the centralized logging system to minimize performance impact.
4. Ensure the implementation respects the logging eligibility conditions https://github.com/woocommerce/woocommerce/issues/49282
5. Implement throttling to control the rate of error logging and prevent system overload.
6. Test the implementation to ensure errors are correctly captured and logged without affecting performance.
7. Filter out errors originating from third parties by dropping error events that exclusively contain third-party stack frames. (Reference: https://github.com/stayallive/wp-sentry?tab=readme-ov-file#capturing-errors-only-from-certain-theme-andor-plugin)

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/49287.

This PR implements server-side (PHP) remote error logging for WooCommerce. This feature is currently disabled by default.

When the feature flag is enabled and meets the criteria, it will capture all Woocomerce-related fatal errors and log them asynchronously to the server. Developer can also use `log()` method to log custom errors.

Some of the key features of this PR are:

- `register_shutdown_function` is used to capture fatal errors (This is how we're currently capturing fatal errors in WooCommerce for status report).
- Sanitize and trace the error message before sending it to the server.
  - ` /var/www/html/wp-content/plugins/woocommerce/includes/class-wc-remote-logger.php on line 123` will be sanitized to: `**/woocommerce/includes/class-wc-remote-logger.php on line 123`
- Set `non-blocking` to `true` in `wp_safe_remote_post` to make the request non-blocking.
- Implement throttling to control the rate of error logging and prevent system overload.
- Filter out errors originating from third parties by dropping error events that exclusively contain third-party stack frames
- Introduce `woocommerce_remote_logger_formatted_log_data` filter to allow other plugins to modify the log data before sending it to the server.
- Introduce `woocommerce_remote_logging_is_third_party_error` filter to allow other plugins to determine if an error is third-party or not. I think this is useful for sending fatal logs to other Woo plugins.

#### Example log entry:

We log the following data to the server:

- host: The host of the site
- message: The error message
- trace: The error trace
- php_version: The PHP version
- wc_version: The WooCommerce version
- wp_version: The WordPress version
- store_id: The store ID

Screenshot 2024-07-18 at 12 09 32

#### Relevant links for reference:

- Logstash API: fbhepr%2Skers%2Sjcpbz%2Schoyvp.ncv%2Serfg%2Sjcpbz%2Qwfba%2Qraqcbvagf%2Spynff.jcpbz%2Qwfba%2Qncv%2Qybtfgnfu%2Qraqcbvag.cuc%3Se%3Qnr27n451-og
- WPCOM WooCommerce Logstash handler: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Sjbbpbzzrepr%2Svapyhqrf%2Spynff%2Qjp%2Qybtfgnfu%2Qunaqyre.cuc%3Se%3Q8p4146o2%2396-og

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR is only testable for WooCommerce core team members since it requires access to the logstash server.

1. Create a fresh install of WooCommerce and `Code Snippet` plugin
2. Add the following code snippet to your site to enable remote logging feature:

```php
update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );

add_filter(
	'option_woocommerce_allow_tracking',
	function () {
		return 'yes';
	}
);

add_filter(
	'option_woocommerce_remote_variant_assignment',
	function () {
		return 5;
	}
);
```

3. Trigger a fatal error in WooCommerce by adding the following code to [`load_rest_api()`](https://github.com/woocommerce/woocommerce/blob/ac834f55f75d4e51955bad43115939faa38ec9da/plugins/woocommerce/includes/class-woocommerce.php#L535):

```php
throw new Exception( 'Test fatal error in woocommerce' );
```

4. Refresh the page and check the fatal error is displayed on the page
5. Reverse the changes made in step 3
6. Trigger a fatal error in a third-party plugin by adding the following code to the `load_classes` function in the `wp-content/plugins/code-snippets/php/admin.php` file

```php
		throw new \Exception( 'Code Snippets admin menus are not loaded' );
```

7. Refresh the page and check the fatal error is displayed on the page
8. Check the log server page for the log entry (See test plan in D155228-code to see how to check the log entry)
9. Confirm that woocommerce fatal errors are logged and third-party fatal errors are not logged
10. Please test around the feature and review the code changes 

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Logging/RemoteLogger.php
Comment (moon0326) [inline plugins/woocommerce/src/Logging/RemoteLogger.php]: Is there a reason why we want to extend WC_Log_Handler when we're not implementing the interface method `handle`?  

It seems like classes that extends `WC_Log_Handler` are used in https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-logger.php#L167

I think RemoteLogger is somewhat specialized logger that is called directly in https://github.com/woocommerce/woocommerce/pull/49599/files#diff-97fb927371ee663e98afaf8b8134905a96d4b59c68f6a83c569ec42a6b468a08R405 and might not need to extend/implement [WC_Log_Handler](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/abstracts/abstract-wc-log-handler.php) or [WC_Log_Handler_Interface](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/interfaces/class-wc-log-handler-interface.php)

Comment (chihsuan) [reply plugins/woocommerce/src/Logging/RemoteLogger.php]: Good question! I extended `WC_Log_Handler` because I wanted to reuse `get_backtrace` method from WC_Log_Handler. I could have just copied the method to RemoteLogger but I thought it would be better to reuse it. What do you think?

https://github.com/woocommerce/woocommerce/blob/12b701d35cac2c7b560d1685f485ce8283f4dc85/plugins/woocommerce/includes/abstracts/abstract-wc-log-handler.php#L66-L97

Comment (moon0326) [reply plugins/woocommerce/src/Logging/RemoteLogger.php]: Ah, I see! I don't think there is any harm in extending it, as we'll most likely use RemoteLogger in a specific place.

My comment was purely from a technical point of view.

LGTM!

**Comment by barryhughes** [comment]
> 4. Refresh the page and check the fatal error is displayed on the page
> 5. Reverse the changes made in step 4

Minor thing, should this be *reverse the changes made in step 3?*

File: plugins/woocommerce/includes/class-woocommerce.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-woocommerce.php]: It doesn't seem like the remote logger has been registered, unless I'm missing something, so whenever execution reaches this point things fail with a new exception:

`Alias (Automattic\WooCommerce\Loggers\RemoteLogger) is not being managed by the container or delegates`

Are there some uncommitted changes we need?

**Review by barryhughes — COMMENTED** [review]
One more thought/question:

- `RemoteLogger` is a `WC_Log_Handler`.
- It's possible for [multiple `WC_Log_Handler` instances to be active](https://github.com/woocommerce/woocommerce/blob/9.1.2/plugins/woocommerce/includes/class-wc-logger.php#L167) at any given time.

Instead of selectively instantiating `RemoteLogger`, what if it is always added as a supplementary logger (so long, of course, as the relevant feature flag is enabled)? Noting here that it could still make a reasoned choice about the type of messages it decides to send to LogStash (ie, it can decline to forward anything that is not critical, etc).

In the long-term, this could be great if we see a future where we might want to go beyond capturing fatals.

File: plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php]: - Should we trailing slash this?
- Should we normalize?
- Should we think about non-standard plugin directories?

On the first point, I'm thinking that the current logic will match paths for any plugins with a path of `plugins/woocommerce-*` (there are a huge number of these, if you scan through the slugs registered in the [plugin directory](https://plugins.svn.wordpress.org/), and there are doubtless others from marketplaces like our own and other sources).

On the second point, I just want to ensure we aren't going to have issues on Windows servers, where perhaps we fail to match because of a difference in directory separators (unless perhaps any paths inside the error message have already been normalized).

Lastly, the plugin directory is not always `wp-content/plugins` (and indeed the content directory may not be `wp-content`, though that isn't the level we're working with here).

File: plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php]: Related to an earlier comment, should we think about non-standard plugin paths? For instance, what if `WP_PLUGIN_DIR` is `/wp/shared/modules`?

**Comment by barryhughes** [comment]
Not directly relevant to *this* PR, but I'm curious if `WC_Site_Tracking::is_tracking_enabled()` should be used to determine if tracking has been opted into.

The current logic contained in `RemoteLogger::is_tracking_opted_in()` doesn't for instance check if there is a filter returning false on `woocommerce_apply_tracking` (and, though that hook is deprecated, some users may still depend on it).

File: plugins/woocommerce/includes/class-woocommerce.php
Comment (chihsuan) [reply plugins/woocommerce/includes/class-woocommerce.php]: Sorry. I updated the namespace upstream but forgot to update it in this branch. I'll fix it.

File: plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php]: These are really good points. 👍  I agree that we should handle trailing slashes and normalize paths. I'll see how we can do that.

**Comment by chihsuan** [comment]
Thanks for the review and great feedback! @barryhughes Appreciate it. 🙇  I'll make the changes you suggested and let you know when it's done.  

> Not directly relevant to this PR, but I'm curious if WC_Site_Tracking::is_tracking_enabled() should be used to determine if tracking has been opted into.

> The current logic contained in RemoteLogger::is_tracking_opted_in() doesn't for instance check if there is a filter returning false on woocommerce_apply_tracking (and, though that hook is deprecated, some users may still depend on it).

I searched our code base for the `woocommerce_allow_tracking` option and found that we use it directly in [many places](https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce+get_option%28+%27woocommerce_allow_tracking%27&type=code). 😅 

I'll change it to use `WC_Site_Tracking::is_tracking_enabled()` instead as it does seems more appropriate.  🙂

**Comment by barryhughes** [comment]
> I searched our code base for the woocommerce_allow_tracking option and found that we use it directly in [many places](https://github.com/search?q=repo%3Awoocommerce%2Fwoocommerce+get_option%28+%27woocommerce_allow_tracking%27&type=code). 😅

Yikes, but thank you for flagging ... this is probably something for us to tighten up (outside of this PR, of course).

File: plugins/woocommerce/includes/class-woocommerce.php
Comment (chihsuan) [reply plugins/woocommerce/includes/class-woocommerce.php]: Fixed in e4e3019a02b53ee72a1c159f2ce3c1f66bc7c502.

File: plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php]: Ok, I've updated the code to address these points in 4a2bc4e01bcf0b5994fe9180b1c30c5b95d19537. Thanks again for bringing them up! 🙏

File: plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Loggers/RemoteLogger.php]: Addressed in the latest update 4a2bc4e01bcf0b5994fe9180b1c30c5b95d19537

**Comment by chihsuan** [comment]
Hey @barryhughes, I've addressed the points you've raised. Could you please take another look at the code and let me know if you have any further feedback? Thanks! 🙏 

Regarding `WC_Log_Handler`, I think it's a good idea to add `RemoteLogger` as a supplementary logger. However, I prefer to keep the current implementation for now so that we can make sure we only log fatal errors for our first iteration. We can iterate on this later. What do you think? 🙂

**Comment by barryhughes** [comment]
> I prefer to keep the current implementation for now

Yep, I think that's fine.

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (adrianduffell) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: I'm a bit confused why we do nothing here. In the [interface](https://github.com/woocommerce/woocommerce/blob/4ac3ce099c5e3cd674858fc258bf4c4d899923f6/plugins/woocommerce/includes/interfaces/class-wc-log-handler-interface.php#L16), it looks like `handle` is intended to do the job that `log` currently has.

To comply with interface, perhaps `log` method should be renamed to `handle` and take the first `$timestamp` param (but do nothing with the timestamp) 🤔

Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: We don't actually use it as wc logger handler in the code (at least for our first implementation). As mentioned in previous comment, I extended WC_Log_Handler just because of reusing `get_backtrace` method from `WC_Log_Handler`.

But that sounds like a good idea to rename `log` method to `handle` so make it less confusing. I will update the PR with this change. Thanks for the suggestion!

Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Renamed it here e1989eaa3a59c3f1112782c6a7aed4157298711c

Comment (moon0326) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: I've given it a bit more thought. 

What do you think about extracting `get_backtrace` to a trait (perhaps LogUtils?) and removing the interface? I think we should not try to implement an interface just to use a utility function from an abstract class. It can confuse developers in the future. We're even forced to rename the method :sweat_smile:

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: If this fails, then it is possible an exception will be raised, and we will enter the following catch block. However, what if it fails without an exception being raised? In these cases, `wp_safe_remote_post()` may return a `WP_Error`. Should we look for and log these failures, too?

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: @moon0326  As discussed with Barry above, we will likely add a remote logger as a WC logger handler. So, I think it's good to keep the interface for now.  Let me know if you have different thoughts on this. Thanks! 🙏

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Hmm...you're right! I didn't handle that case because I noticed that when we set `blocking` to `true,` `wp_safe_remote_post()` will not return a WP_Error object when the response is unsuccessful. But I double-checked the code and saw that it would return a WP_Error object when the request fails (e.g., when the URL is invalid or the user has blocked it). So, we should handle this case as well. Thanks for pointing it out!

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-http.php#L123-L127

Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Ok, I've updated the code to handle WP_Error objects in a58b193145fa46b4602efbeece213f6049287162

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Yep: if we see potential for this to expand (in the future) beyond just fatal/critical errors, I think that makes a lot of sense and—especially if we eventually integrate it in the same way as other logging implementations—it would provide a better degree of consistency (3PDs etc seeking to control logging through existing hooks don't have to figure out how this other system works).

That said, we certainly have a considerable amount of freedom to test different strategies and change them later (because much of the work lives in the internal namespace), so also don't treat my thoughts as a blocker of any kind.

Comment (moon0326) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Thanks @chihsuan @barryhughes 👍 

Makes sense 🚀

**Review by moon0326 — COMMENTED** [review]
LGTM and tested well 👍 🚀

I'll let @barryhughes approve it when he's ready.

File: plugins/woocommerce/includes/class-woocommerce.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-woocommerce.php]: This will fail ... we probably want to do:

```suggestion
			$remote_logger = $container->get( RemoteLogger::class );
			$remote_logger->handle( time(), WC_Log_Levels::CRITICAL, $message, $context );
```

Or possibly:

```suggestion
			$container->get( RemoteLogger::class )->handle( time(), WC_Log_Levels::CRITICAL, $message, $context );
```

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: When we call `wc_doing_it_wrong()` then a message/warning is put together with a format looking something like this:

```
<function> was called incorrectly. <message>. This message was added in version <version>.
```

With that in mind, I'm not sure if there is really any need to embed the method name in the message as we're currently doing, since this is taken care of already. If we decide to keep it, though, we should probably also update it to `::handle()`.

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Optionally, could specify the return type here:

```suggestion
	private function sanitize_trace( $trace ): string {
```

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: *Optional* thing to consider: we could add type-hinting.

```suggestion
	protected function is_third_party_error( string $message, array $context ): bool {
```

Though I realize we'd probably also need a cast further up where this method is called (and so, if we're deliberately being loose, that's probably fine).

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: For consideration: instead of implementing these throttling mechanisms here in the logger, might it reduce the amount of code and perhaps also be more efficient to use `WC_Rate_Limiter`?

**Review by barryhughes — CHANGES_REQUESTED** [review]
Left a note about one required change (to the code in `WooCommerce::log_errors()` method), also some optional changes and other observations.

Please note that, in my own testing, I did **not** look at Logstash itself (though I traced through to the dispatch of the non-blocking request that sends the information there). I definitely recommend we test this element before merge.

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: TIL, I din't know about `WC_Rate_Limiter`. I'll look into it. Thanks for the suggestion! 😃

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: I decided to keep it as is for now because WC Logger also calls `wc_doing_it_wrong()` with the method name in the message.  I think it would be better to keep it consistent.

https://github.com/woocommerce/woocommerce/blob/f3883053f3a50689c034450ba28de6b996d8eb7a/plugins/woocommerce/includes/class-wc-logger.php#L168-L171

Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Updated it to `::handle()` in f3883053f3a50689c034450ba28de6b996d8eb7a

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Replaced it with `WC_Rate_Limiter` in 3e76773b791a587817f501e42c55fbd3aab83ad8

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Sounds good. Added type-hinting and updated code where this method is called in 628fcf690ee0df05679a461d2fdf8ecf10a5335c

**Review by adrianduffell — APPROVED** [review]
Nice work, @chihsuan! It tested well for me -- the exception in the WooCommerce plugin logged, but not the external one. My host was `optimistic-gold-serval.jurassic.ninja`. LGTM! 🚀

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Apologies for missing this one during the earlier passes 😬 

I think there is a risk we'll end up with the local logger set to null (and, we'll subsequently try to call a method on null, if for instance we try to use it to record *'Remote logging throttled').*

We probably need either:

```suggestion
		$this->local_logger = null === $logger
			? wc_get_logger()
			: $logger;
```

Or, if preferred:

```suggestion
		if ( null === $logger ) {
			$this->local_logger = wc_get_logger();
		} else {
			$this->local_logger = $logger;
		}
```

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Minor and non-essential change (so feel free to skip):

```suggestion
			WC_Rate_Limiter::set_rate_limit( self::RATE_LIMIT_ID, self::RATE_LIMIT_DELAY );
```

(We import `WC_Rate_Limiter`, so qualification isn't needed.)

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Also feel free to skip (as it shouldn't cause any problems) but, now that we are using strong typing in the signature, we don't need to test if `$message` is a string.

```suggestion
		if ( str_contains( $message, $wc_plugin_dir ) ) {
```

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (barryhughes) [inline plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Again, completely optional change you can feel free to skip:

```suggestion
		/**
		 * Filter to allow other plugins to overwrite the result of the third-party error check for remote logging.
		 *
		 * @since 9.2.0
		 *
		 * @param bool   $is_third_party_error The result of the third-party error check.
		 * @param string $message              The error message.
		 * @param array  $context              The error context.
		 */
		return apply_filters( 'woocommerce_remote_logging_is_third_party_error', true, $message, $context );
```

We don't need to assign the value-to-be-filtered to a variable, since it is basically a constant (but if you prefer to do it this way for readability, totally fine).

**Review by barryhughes — CHANGES_REQUESTED** [review]
Final few things for consideration. All are optional bar one (relating to initialization of the local logger, which we should address). Thanks again!

File: plugins/woocommerce/src/Internal/Logging/RemoteLogger.php
Comment (chihsuan) [reply plugins/woocommerce/src/Internal/Logging/RemoteLogger.php]: Ah, Good catch!

**Comment by chihsuan** [comment]
> Final few things for consideration. All are optional bar one (relating to initialization of the local logger, which we should address). Thanks again!

Thanks for the feedback! @barryhughes  I've made the changes you suggested. 💜

**Review by barryhughes — APPROVED** [review]
Supoib, no other feedback from me :-)

**Pull request merged**
